### PR TITLE
Synchronization: Synonyms - `synonym_type`

### PR DESCRIPTION
**Pull request by joeflack4**
Merging into:
- #93

## Overview
Synchronization: Synonyms - `synonym_type`
- Update: New feature 'synonym_type' column to synonym sync outputs. Contains a combination of Mondo types declared for mapped synonyms as well as those obtained from the source, either declared in the component, or inferred by logic, e.g. mondo#ABBREVIATION. Previous 'synonym_type' column renamed to 'synonym_type_mondo'.

Also synonym sync related:
- Bug fix: Dropped duplicate rows

### Documentation
Was the documentation added/updated under `docs/`?
- [ ] Yes
- [x] No, updates to the docs were not necessary after careful consideration

### QC
Was the full pipeline run before submitting this PR using `sh run.sh make build-mondo-ingest` on this branch (after 
`docker pull obolibrary/odkfull:dev`), and no errors occurred?
- [ ] Yes
- [x] No, there are no functional (code-related) changes to the pipeline in the PR, so no re-run is necessary
   
### New Packages
Were any new *Python packages* added?
- [ ] Yes, [requirements.txt.full](https://github.com/INCATools/ontology-development-kit/blob/master/requirements.txt.full) was updated and an [ODK issue](https://github.com/INCATools/ontology-development-kit/issues/new) submitted
- [x] No

Were any other *non-Python packages* added?
- [ ] Yes, [Dockerfile](https://github.com/INCATools/ontology-development-kit/blob/master/Dockerfile) updated and an [ODK issue](https://github.com/INCATools/ontology-development-kit/issues/new) submitted
- [x] No

### PR Review and Conversations Resolved
Has the PR been sufficiently reviewed by at least 1 team member of the Mondo Technical team and all threads resolved?
- [ ] Yes

## Additional information
- [Google sheet](https://docs.google.com/spreadsheets/d/1ftEZ47-JDe-1Z_fRWyO5NS39qT_-8MZsHXHEKz9aZ9Q/edit?gid=467804832#gid=467804832)

### CODE REVIEW

File: src/sparql/synonyms-scope-type-xref.sparql
Comment (joeflack4) [inline src/sparql/synonyms-scope-type-xref.sparql:1]: ## FYI: Querying synonym types via SPARQL & ramifications
OAK SqlImplementation doesn't support querying axioms. I was able to refactor and re-use the existing query we were using on Mondo, and now it queries both sources and Mondo.

It is possible now to get rid of OAK completely if we want to refactor; it's redundant.

File: src/scripts/sync_synonym.py
Comment (joeflack4) [inline src/scripts/sync_synonym.py:245]: ## Bug fix: Dropped some duplicate rows
I believe there were a very small amount of duplicate rows that were making their way into the output, and this fixes that.

File: src/scripts/sync_synonym.py
Comment (joeflack4) [inline src/scripts/sync_synonym.py:275]: ## TODO: Support reading 2+ `synonym_type` (source & Mondo)
We can, and do, have instances of multiple `synonym_type` in the ROBOT template.

However, one thing I didn't have time to do yet was to handle edge cases of multiple synonym types are when querying the axioms from Mondo or sources.

For now, I've set up an error to alert when this happens. I got no error, so there are no current instances. However, there will be in the future. That's why I've marked this with an uppercase `TODO`; I will handle it soon.

File: src/scripts/sync_synonym.py
Comment (joeflack4) [inline src/scripts/sync_synonym.py:75]: ## Relationship of `synonym_type` to `synonym_type_mondo`
- [ ] @matentzn @twhetzel I think you guys may want to pay more attention to this topic. Want to make sure what I did was correct.

`synonym_type_mondo` shows all synonym types currently declared on the synonym from Mondo.

`synonym_type` is used to import new synonyms into Mondo.
However, I was worried. I didn't get time to check, but I think that what is in this column will potentially overwrite what's already in Mondo.

For example, say that a synonym in Mondo has `syn_type_a`, but the synonym sync finds that it also has a new type `syn_type_b`. I think that if the `synonym_type` field is only set to 'b', then 'a' will get overwritten. So what I've done is merged them. Now in this case, `synonym_type` gets set to `syn_type_a|syn_type_b`.
I haven't had a chance yet to test this 'overwritten' behavior, but if I am wrong, I can change this to not merge the two.

<details><summary>Additional granular synonym_type fields I considered</summary>
<p>

I'm not a curator so I can't say, but it might be useful for us to be able to distinguish the following cases:
- The synonym type from Mondo
- Declared: The synonym type declared in the source artefact
  - e.g. OMO:0003012 is declared in the DOID mirror
- Inferred (via synonym sync): New synonym type generated through the synonym sync logic
  - e.g. <10 chars, all caps, etc = acronym
- Created through a non-synonym-sync pipeline: e.g. `MONDO:GENERATED`, which is on several components. I think that the component creation pipeline makes some of these, but some of them also might get made during individual source pipelines (e.g. the OMIM pipeline in its own git repo)

</p>
</details>

File: src/scripts/sync_synonym.py
Comment (joeflack4) [inline src/scripts/sync_synonym.py:131]: ## Language code removal
- [ ] @twhetzel @matentzn Review

### What this is doing
> `pattern`: Default will capture either of these cases, 
> (1) String ends with @ followed by 2-3 characters,
> (2) String ends with @ followed by 2-3 characters, followed by a -, followed by an number of characters.

> Complies with RDF & OWL's language tag formatting based on IETF's BCP 47 (Best Current Practice 47) standard.

### Why a change like this was necessary
I couldn't use OAK to query synonym types, so I used SPARQL. But unlike with OAK (or SemSQL), many synonyms are coming through with `@en`, some `@fr`, and possibly others.  I think 3-4 sources had these (DO, Orphanet, and ICD10CM at least I think).

Using Claude I was easily able to make a regex meeting those conditions above, which handles simple cases, e.g. `@fr`, but also more complex cases like `@zh-Hans-CN`. I did some quick true/false negative/positive testing, and it passed my checks.

### What I'd like to do in the future
Find an existing library for this, or otherwise implement this in `mondolib`.

### What I'd like from review
Just wanna know if you guys feel dubious about the regex. If so, then maybe for now we can go with a simple check for `@en` and `@fr`. However, I don't know if this will 100% cover all cases for our current sources without further analysis.

File: src/ontology/reports/icd10who.synonyms.confirmed.robot.tsv
Comment (matentzn) [inline src/ontology/reports/icd10who.synonyms.confirmed.robot.tsv:3]: CLINGEN_LABEL is only provenance. It should absolutely not appear in an active ROBOT template column!

image

File: src/ontology/reports/icd10who.synonyms.updated.robot.tsv
Comment (matentzn) [inline src/ontology/reports/icd10who.synonyms.updated.robot.tsv:1]: - `synomym` column is a bit redundant (if you wanted to make space), no biggie if not
- Did you make sure that robot correctly adds the evidence even if the synonym column is not directly preceding the evidence column?

File: src/scripts/sync_synonym.py
Comment (matentzn) [inline src/scripts/sync_synonym.py:1]: I did not review this file

**Review by matentzn — CHANGES_REQUESTED** [review]
One major issue, some minor I could see!

File: src/ontology/mondo-ingest.Makefile
Comment (twhetzel) [reply src/ontology/mondo-ingest.Makefile:605]: There was no discussion about `$(REPORTDIR)/%-synonyms.deleted.robot.tsv`.

File: src/ontology/mondo-ingest.Makefile
Comment (twhetzel) [inline src/ontology/mondo-ingest.Makefile:645]: nit: If `onto-synonyms-path` is for synonym scope types, be more specific in the naming of `onto-synonyms-path`.

File: src/ontology/reports/icd10who.synonyms.confirmed.robot.tsv
Comment (twhetzel) [reply src/ontology/reports/icd10who.synonyms.confirmed.robot.tsv:3]: Agreed. This also occurs on Line 13434 where the source is DOID and perhaps other lines. 

The synonym type `CLINGEN_LABEL` is only a synonym type in Mondo so it's not clear to me why it is showing up in the column `synonym_type`, which I understand should only have source synonym type information.

File: src/scripts/sync_synonym.py
Comment (twhetzel) [reply src/scripts/sync_synonym.py:69]: Abbreviations can have numbers, e.g. MONDO:0007182 Machado-Joseph disease with synonym SCA3.

This line in the v5 Google sheet should not have MONDO#ABBREVIATION in the `synonym_type` column. DOID already tags this synonym with `OMO_0003012` and the script should maintain that synonym type provenance.  I've posted in Slack that we will be discussing how to handle these type of cases for DOID in the Tech Call.
<google-sheets-html-origin><style type="text/css"></style>
MONDO:0007182 | Machado-Joseph disease | oio:hasExactSynonym |   | SCA3 | SCA3 |   |   |   | DOID:1440 | Machado-Joseph disease | http://purl.obolibrary.org/obo/mondo#ABBREVIATION | http://purl.obolibrary.org/obo/mondo#ABBREVIATION |   | confirmed
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --

File: src/scripts/sync_synonym.py
Comment (twhetzel) [reply src/scripts/sync_synonym.py:75]: The source synonym type information and mondo synonym type information should not be merged into the `synonym_type` column. This column should only contain any source synonym type information that already exists in the source such as OMO:0003012 _or_ the inferred synonym type based on the abbreviation rule in the script.

File: src/scripts/sync_synonym.py
Comment (twhetzel) [reply src/scripts/sync_synonym.py:131]: This (removing any language tags like `@en`) should be addressed in the SPARQL query.

File: src/scripts/sync_synonym.py
Comment (twhetzel) [reply src/scripts/sync_synonym.py:249]: What are some examples of `Filter out synonyms not owned by source, but in source artefact (e.g. ChEBI)`?

At this time, `OMO:0003012` should not be converted to `MONDO:ABBREVIATION`. The original source synonym type information should be in the google sheet.

File: src/scripts/sync_synonym.py
Comment (twhetzel) [reply src/scripts/sync_synonym.py:275]: Comment out this section for now until we can discuss how this should be handled.

**Review by twhetzel — CHANGES_REQUESTED** [review]
See inline comments.

File: src/scripts/sync_synonym.py
Comment (joeflack4) [reply src/scripts/sync_synonym.py:69]: Yeah IDK what I was thinking w/ the `.isalpha()` now that I think of it. Will take a look at slack.

File: src/scripts/sync_synonym.py
Comment (joeflack4) [reply src/scripts/sync_synonym.py:75]: Sure, I'll change that.

I assume that means that we do not think that the ROBOT behavior here will be to overwrite the synonym type in Mondo, but to append to it.

File: src/scripts/sync_synonym.py
Comment (joeflack4) [reply src/scripts/sync_synonym.py:249]: Oh, I thought you asked for that. OK I will undo that.

File: src/scripts/sync_synonym.py
Comment (joeflack4) [reply src/scripts/sync_synonym.py:131]: I was thinking of that but when this has come up before we've done it in Python. I'll look into how to do that in the SPARQL query.

I think probably this change will need go into next week's changs.

File: src/ontology/mondo-ingest.Makefile
Comment (joeflack4) [reply src/ontology/mondo-ingest.Makefile:605]: @matentzn We all decided we are not doing the 'deleted' case yet. It was more complicated than we originally imagined.

That `tmp/` file is the google sheet everything concatenated together. It's the google sheet we are using. If that shouldn't be a tmp file, let me know. If it should be, then this is here just temporarily (there is a `todo` comment above it about that) so that it's easier for me to run.

File: src/ontology/reports/icd10who.synonyms.confirmed.robot.tsv
Comment (joeflack4) [reply src/ontology/reports/icd10who.synonyms.confirmed.robot.tsv:3]: This has to do with the merging of synonym_type and synonym_type_mondo. I have another comment about that.

But it looks like that concern is not founded, so I will not merge these and then this will go away.

File: src/ontology/reports/icd10who.synonyms.updated.robot.tsv
Comment (joeflack4) [reply src/ontology/reports/icd10who.synonyms.updated.robot.tsv:1]: I didn't get to check that yet.

As discussed, the synonym and synonym_scope columns are here to make curation easier but if you think that is not the case I can remove.

File: src/sparql/synonyms-scope-type-xref.sparql
Comment (joeflack4) [reply src/sparql/synonyms-scope-type-xref.sparql:1]: We may do the OAK to SPARQL refactor in the future. Not a high priority.

File: src/scripts/sync_synonym.py
Comment (joeflack4) [reply src/scripts/sync_synonym.py:275]: Discussed and we will prioritize this soon; likely next week.

File: src/ontology/mondo-ingest.Makefile
Comment (joeflack4) [reply src/ontology/mondo-ingest.Makefile:621]: This file is what gets uploaded as the [google sheet](https://docs.google.com/spreadsheets/d/1ftEZ47-JDe-1Z_fRWyO5NS39qT_-8MZsHXHEKz9aZ9Q/edit?gid=467804832#gid=467804832).

Trish asked to move this to `reports/` instead of `tmp/`.

It's unclear now, but it may be that this is something we want to keep permanently.

File: src/scripts/sync_synonym.py
Comment (joeflack4) [reply src/scripts/sync_synonym.py:249]: Also the filtering of CheBi we decided to continue doing in Python.

File: src/ontology/mondo-ingest.Makefile
Comment (joeflack4) [reply src/ontology/mondo-ingest.Makefile:645]: WIll change

**Comment by joeflack4** [comment]
Reviewed with Trish and approved. We are merging this feature branch into another feature branch. There are still a couple outstanding issues brought up here, but those will be handled in:
- #93

**Comment by twhetzel** [comment]
Joe, we reviewed this PR however I did not officially approve the PR or see the final code updates. This is a bad working process to say someone approved a PR and merge it when it was not approved.

**Comment by twhetzel** [comment]
Generally looks good. I saw a note the files were re-generated, but still directly under `reports` vs. `reports/sync-synonyms`.

**Comment by joeflack4** [comment]
@twhetzel Oh, sorry Trish. I asked if I would merge this PR once I pushed these new changes, and it sounded like you said yes.

Also this is a feature --> feature, so there's still a lot of room to tweak things. If you want, take a look at the [commits](https://github.com/monarch-initiative/mondo-ingest/pull/639/commits) I made, and if you add comments/requests on those diffs I will act on them.

`reports/syncs-synonyms/` I have in my personal notes for next week, but I've just added a [tracker comment](https://github.com/monarch-initiative/mondo-ingest/pull/93#discussion_r1719431883); will reply there when done.

**Pull request merged**
